api/query/cache/service/main.go

```diff
@@ -67,166 +63,12 @@ func readinessCheckHandler(w http.ResponseWriter, r *http.Request) {
-			ids = append(ids, rq.RunIDs[i])
-			runs = append(runs, run)
-		}
-	}
-
-	// Return to client `http.StatusUnprocessableEntity` immediately if any runs
-	// are missing.
-	if len(runs) == 0 && len(missing) > 0 {
-		data, err = json.Marshal(shared.SearchResponse{
-			IgnoredRuns: missing,
-		})
-		if err != nil {
-			log.Errorf("Failed to marshal results: %s", err.Error())
-			http.Error(w, "Failed to marshal results to JSON", http.StatusInternalServerError)
-		}
-		w.WriteHeader(http.StatusUnprocessableEntity)
-		w.Write(data)
-		return
-	}
-
-	// Prepare user query based on `ids` that are (or at least were a moment ago)
-	// resident in `idx`. In the unlikely event that a run in `ids`/`runs` is no
-	// longer in `idx`, `idx.Bind()` below will return an error.
-	q := cq.PrepareUserQuery(ids, rq.AbstractQuery.BindToRuns(runs...))
-
-	// Configure format, from request params.
-	urlQuery := r.URL.Query()
-	subtests, _ := shared.ParseBooleanParam(urlQuery, "subtests")
-	interop, _ := shared.ParseBooleanParam(urlQuery, "interop")
-	diff, _ := shared.ParseBooleanParam(urlQuery, "diff")
-	diffFilter, _, err := shared.ParseDiffFilterParams(urlQuery)
-	if err != nil {
-		log.Errorf("%s", err.Error())
-		http.Error(w, err.Error(), http.StatusBadRequest)
-		return
-	}
-	opts := query.AggregationOpts{
-		IncludeSubtests:         subtests != nil && *subtests,
-		InteropFormat:           interop != nil && *interop,
-		IncludeDiff:             diff != nil && *diff,
-		DiffFilter:              diffFilter,
-		IgnoreTestHarnessResult: shared.IsFeatureEnabled(store, "ignoreHarnessInTotal"),
-	}
-	plan, err := idx.Bind(runs, q)
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-
-	results := plan.Execute(runs, opts)
-	res, ok := results.([]shared.SearchResult)
-	if !ok {
-		log.Errorf("Search index returned bad results: %s", err.Error())
-		http.Error(w, "Search index returned bad results", http.StatusInternalServerError)
-		return
-	}
-
-	// Cull unchanged diffs, if applicable.
-	if opts.IncludeDiff && !opts.DiffFilter.Unchanged {
-		for i := range res {
-			if res[i].Diff.IsEmpty() {
-				res[i].Diff = nil
-			}
-		}
-	}
-
-	// Response always contains Runs and Results. If some runs are missing, then:
-	// - Add missing runs to IgnoredRuns;
-	// - (If no other error occurs) return `http.StatusUnprocessableEntity` to
-	//   client.
-	resp := shared.SearchResponse{
-		Runs:    runs,
-		Results: res,
-	}
-	if len(missing) != 0 {
-		resp.IgnoredRuns = missing
-	}
-
-	data, err = json.Marshal(resp)
+	err := searchHandlerImpl(w, r)
 	if err != nil {
-		log.Errorf("Failed to marshal results: %s", err.Error())
-		http.Error(w, "Failed to marshal results to JSON", http.StatusInternalServerError)
-		return
-	}
-	if len(missing) != 0 {
-		w.WriteHeader(http.StatusUnprocessableEntity)
+		log := shared.GetLogger(r.Context())
+		log.Errorf(err.Error())
+		http.Error(w, err.Message, err.Code)
```

Comment:
My concern here is that because `w` has been passed into `searchHandlerImpl`, we have no guarantee that `w` has not already been written to. What happens if you call `http.Error` after calling (say) `w.Write(...)` - does it panic?

Reply:
No, it does not. It'd fail to update the status code; technically it's a behaviour not defined in Golang docs.